### PR TITLE
[UNDERTOW-1917] Logic in try/finally blocks hides exceptions

### PR DESCRIPTION
**Pull request by boris-unckel**
Fixes https://issues.redhat.com/browse/UNDERTOW-1917

### CODE REVIEW

**Comment by boris-unckel** [comment]
The solution is inspired by @jfdenise proposal, already working in WFCore: [https://github.com/wildfly/wildfly-core/pull/4615](https://github.com/wildfly/wildfly-core/pull/4615)

**Comment by boris-unckel** [comment]
@fl4via Hello Flavia, I have prepared this PR to fix a bug regarding exception suppressing. Please, could you review and advise what is needed to polish it? After everything is prepared for the actual version I can provide the backport also, if you want to have this fixed for older versions, too.

**Review by fl4via — APPROVED** [review]

File: core/src/main/java/io/undertow/protocols/http2/Http2PushBackParser.java [hunk 1]
```diff
@@ -63,28 +64,48 @@ public void parse(ByteBuffer data, Http2FrameHeaderParser headerParser) throws I
...
+                } else {
+                    original = t;
+                }
             }
-            data.limit(oldLimit);
-            remainingData -= used;
-            if (remainingData == 0) {
-                finished = true;
+            if (original != null) {
+                if (original instanceof RuntimeException) {
```
Comment (ropalka) [inline core/src/main/java/io/undertow/protocols/http2/Http2PushBackParser.java:100]: All three 'instanceof' checks are useless .

File: core/src/main/java/io/undertow/security/impl/FormAuthenticationMechanism.java [hunk 2]
```diff
@@ -129,16 +132,35 @@ public AuthenticationMechanismOutcome runFormAuth(final HttpServerExchange excha
...
+                    }
                 }
-                return outcome != null ? outcome : AuthenticationMechanismOutcome.NOT_AUTHENTICATED;
             }
         } catch (IOException e) {
-            throw new RuntimeException(e);
+            original = new UncheckedIOException(e);
+        }
+        if (original != null) {
+            if (original instanceof RuntimeException) {
```
Comment (ropalka) [inline core/src/main/java/io/undertow/security/impl/FormAuthenticationMechanism.java:156]: Both 'instanceof' checks are useless.

File: core/src/main/java/io/undertow/server/protocol/ajp/AjpServerRequestConduit.java [hunk 3]
```diff
@@ -320,20 +322,47 @@ private int doRead(final ByteBuffer dst, long state) throws IOException {
...
+                }
+            } catch (Throwable t) {
+                if (suppressed != null) {
+                    suppressed.addSuppressed(t);
+                } else {
+                    suppressed = t;
+                }
+            }
+            if (suppressed != null) {
+                if (suppressed instanceof RuntimeException) {
```
Comment (ropalka) [inline core/src/main/java/io/undertow/server/protocol/ajp/AjpServerRequestConduit.java:354]: All three 'instanceof' checks are useless.

**Review by ropalka — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/undertow/protocols/http2/Http2PushBackParser.java (hunk 1 above)
Comment (boris-unckel) [reply core/src/main/java/io/undertow/protocols/http2/Http2PushBackParser.java:100]: @ropalka Could you explain the issue, please? The compiler refuses to throw the (undeclared) Throwable

Comment (ropalka) [reply core/src/main/java/io/undertow/protocols/http2/Http2PushBackParser.java:100]: You're right of course!

**Review by ropalka — APPROVED** [review]
LGTM

**Pull request merged**